cartridges/int_algolia/cartridge/scripts/services/algoliaServiceHelper.js

```diff
@@ -143,6 +141,131 @@ function callJsonService(title, service, params) {
     return callStatus;
 }
 
-module.exports.callService = callService;
-module.exports.callJsonService = callJsonService;
-module.exports.UNEXPECTED_ERROR_CODE = UNEXPECTED_ERROR_CODE;
+/**
+ * Replaces wildcard `*` with `.*` and does a full ^...$ match
+ * @param {string} indexName - e.g. "test__products__en_US"
+ * @param {string} pattern   - e.g. "test*" or "prod_*" or "*_dev"
+ * @returns {boolean} whether indexName matches that pattern
+ */
+function matchesIndexName(indexName, pattern) {
+    // Escape regex special chars except '*'
+    var escaped = pattern.replace(/[.+?^${}()|[\]\\]/g, '\\$&');
+    // Replace '*' with '.*'
+    escaped = escaped.replace(/\*/g, '.*');
+    var re = new RegExp('^' + escaped + '$');
+    return re.test(indexName);
+}
+
+/**
+ * Validate an Algolia API Key's ACLs and prefix coverage.
+ *
+ * The key can be restricted to certain sets of indices by "indexes": [...]. If that field is missing
+ * or empty, the key isn't restricted by index name, so only ACL checks matter. Otherwise, we check
+ * if at least one of the "indexes" patterns covers the user-entered indexPrefix.
+ *
+ * @param {dw.svc.Service} service
+ * @param {string} applicationID
+ * @param {string} apiKey
+ * @param {string} indexPrefix         - the user-provided prefix in BM
+ * @param {boolean} isAdminKey
+ * @returns {Object} { error: Boolean, errorMessage: String, warning: String }
+ */
+function validateAPIKey(service, applicationID, apiKey, indexPrefix, isAdminKey) {
+    // 0) Build the required ACL array
+    var requiredACLs;
+    if (isAdminKey) {
+        requiredACLs = ['addObject', 'deleteObject', 'deleteIndex', 'settings'];
+    } else {
+        // minimal read usage is  "search", (initial values for search : "search", "listIndexes", "settings)
+        requiredACLs = ['search'];
+    }
+
+    // 1) Retrieve key info from Algolia
+    var keyResponse = service.call({
+        method: 'GET',
+        url: 'https://' + applicationID + '.algolia.net/1/keys/' + apiKey
+    });
+
+    if (!keyResponse.ok) {
+        return {
+            error: true,
+            errorMessage: Resource.msg('algolia.error.key.validation', 'algolia', null)
+        };
+    }
+
+    var keyData = keyResponse.object.body;
+    var actualACLs = keyData.acl || [];
+
+    // 2) Identify missing vs. excessive ACLs
+    var missingACLs = requiredACLs.slice();
+    var excessiveACLs = [];
+    actualACLs.forEach(function (acl) {
+        var idx = missingACLs.indexOf(acl);
```

Comment:
Very confusing error prone: here `missingACLs` actually contains the `requiredACLs`
Sometimes it's better to optimize less but have more maintainable code, like doing one `forEach` to get the missingACLs and one other to get the excessiveACLs